service/storage_service.cc

```diff
@@ -830,6 +830,84 @@ void storage_service::scan_cdc_generations() {
     }
 }
 
+future<> storage_service::check_and_repair_cdc_streams() {
+    return async([this] { 
+        auto latest = _cdc_streams_ts;
+        const auto& endpoint_states = _gossiper.get_endpoint_states();
+        for (const auto& [addr, state] : endpoint_states) {
+            if (!_gossiper.is_normal(addr))  {
+                throw std::runtime_error(format("All nodes must be in NORMAL state while performing check_and_repair_cdc_streams"
+                        " ({} is in state {})", addr, _gossiper.get_gossip_status(state)));
+            }
+
+            const auto ts = cdc::get_streams_timestamp_for(addr, _gossiper);
+            if (!latest || (ts && *ts > *latest)) {
+                latest = ts;
+            }
+        }
+
+        bool should_regenerate = false;
+        std::optional<cdc::topology_description> gen;
+        try {
+            gen = _sys_dist_ks.local().read_cdc_topology_description(
+                    *latest, { _token_metadata.count_normal_token_owners() }).get0();
+        } catch (exceptions::request_timeout_exception& e) {
+            throw; // Stop and inform the API caller about timeout
+        } catch (exceptions::unavailable_exception& e) {
+            throw; // Stop and inform the API caller about unavailability
+        } catch (...) {
+            const auto ep = std::current_exception();
+            if (is_timeout_exception(ep)) {
+                throw; // Stop and inform the API caller about other timeout errors
+            }
+            // On exotic errors proceed with regeneration
+            cdc_log.error("Exception while reading CDC topology description: \"{}\". Regenerating streams anyway.", ep);
+            should_regenerate = true;
+        }
+
+        if (!gen) {
+            cdc_log.error(
+                "Could not find CDC generation with timestamp {} in distributed system tables (current time: {}),"
+                " even though some node gossiped about it.",
+                latest, db_clock::now());
+            should_regenerate = true;
+        } else {
+            std::unordered_set<dht::token> gen_ends;
+            for (const auto& entry : gen->entries()) {
+                gen_ends.insert(entry.token_range_end);
+            }
+            for (const auto& metadata_token : _token_metadata.sorted_tokens()) {
+                if (!gen_ends.count(metadata_token)) {
+                    should_regenerate = true;
+                    break;
+                }
+            }
+        }
+
+        if (!should_regenerate) {
+            if (latest != _cdc_streams_ts) {
+                do_handle_cdc_generation(*latest);
+            }
+            cdc_log.info("CDC generation does not need repair");
```

Comment:
Optional: I would add here the generation's timestamp.
`cdc_log.info("CDC generation {} does not need repair", latest)`.

Reply:
done